include/RAJA/exec-openmp/target_reduce_openmp.hxx

```diff
@@ -0,0 +1,498 @@
+    parent(other.parent ? other.parent : &other),
+    val(other.val),
+    idx(other.idx)
+  {
+  }
+
+  //
+  // Destruction releases the shared memory block chunk for reduction id
+  // and id itself for others to use.
+  //
+  ~ReduceMaxLoc<omp_target_reduce, T>()
+  {
+    if (parent) {
+#pragma omp critical
+      {
+        parent->maxloc(val, idx);
+      }
+    }
+  }
+
+  //
+  // Operator that returns reduced max value.
+  //
+  operator T()
+  {
+    return val;
+  }
+
+  //
+  // Method that returns reduced max value.
+  //
+  T get() { return operator T(); }
+
+  //
+  // Method that returns index corresponding to reduced max value.
+  //
+  Index_type getLoc()
+  {
+    return idx;
+  }
+
+  //
+  // Method that updates max and index values for current thread.
+  //
+  const ReduceMaxLoc<omp_target_reduce, T>& maxloc(T rhs, Index_type rhs_idx) const
+  {
+    if (rhs > val) {
+      val = rhs;
+      idx = rhs_idx;
+    }
+    return *this;
+  }
+
+  ReduceMaxLoc<omp_target_reduce, T>& maxloc(T rhs, Index_type rhs_idx)
+  {
+    if (rhs > val) {
+      val = rhs;
+      idx = rhs_idx;
+    }
+    return *this;
+  }
+
+private:
+  //
+  // Default ctor is declared private and not implemented.
+  //
+  ReduceMaxLoc<omp_target_reduce, T>();
+
+  const my_type * parent;
+
+  mutable T val;
+  mutable Index_type idx;
+};
+
+/*!
+ ******************************************************************************
+ *
+ * \brief  Sum reducer class template for use in OpenMP execution.
+ *
+ *         For usage example, see reducers.hxx.
+ *
+ ******************************************************************************
+ */
+
+template <typename T>
+class ReduceSum<omp_target_reduce, T>
+{
+  using my_type = ReduceSum<omp_target_reduce, T>;
+
+public:
+  //
+  // Constructor takes default value (default ctor is disabled).
+  //
+  explicit ReduceSum(T init_val, T initializer = 0)
+    : parent(NULL), val(init_val), custom_init(initializer)
+  {
+	hostid = omp_get_initial_device();
+	devid  = omp_get_default_device();
+	dev_val = new T;
+	is_mapped = false;
+	my_lev = 0;
+        dev_val = (T *)omp_target_alloc( sizeof(T), devid );
+	if( dev_val == NULL ){
+		printf("Unable to allocate space on device\n" );
+		exit(1);
+	}
+	omp_target_memcpy( (void *)dev_val, (void *)&val, sizeof(T), 0, 0, devid, hostid );
+  }
+
+  //
+  // Copy ctor.
+  //
+  ReduceSum(const ReduceSum<omp_target_reduce, T>& other) :
+    parent(other.parent ? other.parent : &other), 
+    val(other.custom_init),
+    dev_val(other.dev_val),
+    custom_init(other.custom_init),
+    is_mapped(other.is_mapped), my_lev(other.my_lev+1), hostid(other.hostid), devid(other.devid)
+  {
+  }
+
+  //
+  // Destruction releases the shared memory block chunk for reduction id
+  // and id itself for others to use.
+  //
+  ~ReduceSum<omp_target_reduce, T>()
+  {
+
+   if( omp_in_parallel() && !omp_is_initial_device() )
+    { 
+#pragma omp critical
```

Comment:
Since the current foreach just does target + parallel for, this should work, but it will break as soon as teams and distribute go in.  The `atomic` directives are guaranteed to work for types up to 64bit though, so that might be a good solution.